drivers/can/can_net.c

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright (c) 2019 Alexander Wachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <net/can.h>
+#include <net/net_pkt.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_can, CONFIG_CAN_NET_LOG_LEVEL);
+
+#define CAN_NET_FILTER_NOT_SET -1
+
+struct net_can_context {
+	struct device *can_dev;
+	struct net_if *iface;
+	int rcv_filter_id;
+	int mcast_filter_id;
+};
+
+static inline u8_t get_frame_datalength(struct zcan_frame *frame)
+{
+	/* TODO: Needs update when CAN FD support is added*/
+	return frame->dlc;
+}
+
+static inline u16_t get_can_lladdr_src(struct zcan_frame *frame)
+{
+	return (frame->ext_id >> CAN_NET_IF_ADDR_SRC_POS) &
+		     CAN_NET_IF_ADDR_MASK;
+}
+
+static inline u16_t get_can_lladdr_dest(struct zcan_frame *frame)
+{
+	return (frame->ext_id >> CAN_NET_IF_ADDR_DEST_POS) &
+		     CAN_NET_IF_ADDR_MASK;
+}
+
+static inline void set_lladdr(struct net_pkt *pkt, struct zcan_frame *frame)
+{
+	struct net_buf *buf = pkt->buffer;
+
+	net_pkt_lladdr_src(pkt)->addr = buf->data;
+	net_pkt_lladdr_src(pkt)->len = sizeof(struct net_can_lladdr);
+	net_pkt_lladdr_src(pkt)->type = NET_LINK_CANBUS;
+	net_buf_add_be16(buf, get_can_lladdr_src(frame));
+	net_buf_pull(buf, sizeof(u16_t));
```

Comment:
Why net_buf_pull() here, the data before the data pointer is now inaccesible?

Reply:
This bytes are the lladdr now. they are referenced by the pkt. The rest of the data is the data in the frame then.

Reply:
> This bytes are the lladdr now. they are referenced by the pkt. The rest of the data is the data in the frame then.

Yes, I understood that. It is just that all of our APIs kind of do not see that any more because the buf->data pointer is moved forward. Not sure if this is an issue thou.

Reply:
I did this because net_pkt_cursor_init sets the cursor to the actual data then instead of the  link layer address. Otherwise i have to skip the first 4 bytes everywhere else in the code.

Reply:
@jukkar This indeed needs some rework. @tbursztyka pointed me to a better solution.

Reply:
@jukkar I tried to remove it here but it seems to make difficulties in canbus.c.
The problem is that there are several situations where the first bytes would have to be skipped. IMO it's better to just pull them here.

Reply:
> IMO it's better to just pull them here.

Ok.